### ISSUE TITLE
A more reliable way to manage the backlight restoring

### ISSUE THREAD

**Issue by Digitalone1** [systemd/systemd#16630]
**systemd version the issue has been seen with**
> 245.7-1

**Used distribution**
 > Arch Linux
Kernel 5.7.10 and latest `mesa` and `xf86-video-amdgpu`.

**VGA and backlight driver**
```
$ lspci
00:01.0 VGA compatible controller: Advanced Micro Devices, Inc. [AMD/ATI] Mullins [Radeon R3 Graphics] (rev 40)

$ ls /sys/class/backlight/
amdgpu_bl0
```

**Expected behaviour you didn't see**
Backlight restored at system startup.

**Unexpected behaviour you saw**
Backlight NOT restored at system startup, even if the related service is showing no errors.

```
$ systemctl status systemd-backlight@backlight:amdgpu_bl0.service
● systemd-backlight@backlight:amdgpu_bl0.service - Load/Save Screen Backlight Brightness of backlight:amdgpu_bl0
     Loaded: loaded (/usr/lib/systemd/system/systemd-backlight@.service; static; vendor preset: disabled)
     Active: active (exited) since Fri 2020-07-31 08:26:17 CEST; 8min ago
       Docs: man:systemd-backlight@.service(8)
    Process: 277 ExecStart=/usr/lib/systemd/systemd-backlight load backlight:amdgpu_bl0 (code=exited, status=0/SUCCESS)
   Main PID: 277 (code=exited, status=0/SUCCESS)
```

**Steps to reproduce the problem**
As described [here](https://bbs.archlinux.org/viewtopic.php?id=257810), just install these packages:

```
systemd-libs (245.6-8 -> 245.7-1)
systemd (245.6-8 -> 245.7-1)
systemd-sysvcompat (245.6-8 -> 245.7-1)
```

The issue is resolved downgrading to previous **245.6-8** version.

I already had this issue in the past months, but I thought it was the kernel, indeed using the lts version resolved the problem. This time it's working with 5.7.10 and it's the systemd upgrade which breaks backlight restoring.

If I disable the service with `systemd.restore_state=0` kernel boot parameter and set up [brillo](https://gitlab.com/cameronnemo/brillo), it's working. In the past I also tried uninstalling `xf86-video-amdgpu` and brillo was restoring the backlight correctly at startup, while systemd service was failing.

It's clear that brillo has a more reliable way to manage the backlight restoring considering that it's not even dependent on X11 packages. Can systemd have the same functionality? Thanks.

**Comment by yuwata** [comment]
Duplicate of #16600.

**Comment by Digitalone1** [comment]
Well, a duplicate, but not totally. If the kernel broke something, why was systemd working in the previous version?

I'd like to focus on a different method to manage this issue since brillo is handling it in a better way while systemd keeps breaking. That happened in the past also.

**Comment by DanielFa** [comment]
Systemd changed behavior in this pull request: https://github.com/systemd/systemd/pull/16316 , which got into the build with 245.7-1 I guess. Since then, actual_brightness is used instead of brightness. Since amd changed the resolution of actual_brightness to 16 bit, but left max_brightness and brightness as 8 bit, the service sees that the stored value is greater than max_brightness and just sets the max brightness. 
Maybe a range check, to check if the read value from actual_brightness is within [0;max_brightness] would be in order, until the kernel devs fix the underlying issue? I know its not directly systemd fault, but it starts to affect more and more people with this update. If actual_brightness is out of range, a fallback to brightness could be done, and if max_brightness gets updated to 16 bit OR the change for actual_brightness is reverted, the fix would continue to work until it gets removed, since it would be well within range then.

I do not have enough experience with systemd for now to submit a merge request, but maybe I will get to that in the coming days, or maybe someone else will do it, unless the won't fix stands, since it is really not a systemd problem, but it could be mitigated until a more permanent fix is in place.

**Comment by yuwata** [comment]
I think only we can do on systemd side is introducing a mechanism to refuse to use `actual_brightness`. For example, introduce `ID_BACKLIGHT_USE_BRIGHTNESS=` udev boolean property and set it for amdgpu.

But still that is a workaround for kernel issue https://bugzilla.kernel.org/show_bug.cgi?id=203905. We should wait for it is fixed on kernel side.

**Comment by poettering** [comment]
Yes, v246 works differently than v245, we use a new kernel API we didn't use before. Turns out on some driver that new API is slightly borked. But it should be fixed in that driver and not worked around in systemd. Please, just work with the relevant kernel devs to fix this.

**Comment by yuwata** [comment]
(Note that PR #16316 is backported to v245 stable. So v245 users also affect this, as the original reporter mentioned so.)

**Comment by Digitalone1** [comment]
I correct myself. Now uninstalling `x86-video-amdgpu` is not disabling the backlight restoring like it was happening in the previous versions.

About `max_brightness` issue, avoiding its check and restoring only the `actual_brightness` would be a fix? I think that it's what brillo is doing and it was working when I was using it.

**Comment by Digitalone1** [comment]
A patch that fixes this issue has been applied, but not released yet in the latest kernel.

https://lore.kernel.org/lkml/a5182129-27fa-d0ff-8aa5-a125aaaf8408@amd.com/

**Comment by poettering** [comment]
OK, let's close this then, as this apparently has been addressed upstream in the kernel now, where it should be addressed. Thanks for posting this upstream!
